### PR TITLE
[Feature] Elemental Task w/ mediaconvert & medialive monthly/daily ingest in ES

### PR DESCRIPTION
**Pull request by jgchoppe**
This PR is about the new task "elemental-process-account". This task ingest in the ElasticSearch data about Elemental Products. In this PR there is also the Elemental MediaConvert & MediaLive monthly/daily ingest.

### CODE REVIEW

**Comment by Dr0p42** [comment]
@jgchoppe where are we about this pull request? Is it still in progress?

File: server/taskProcessAccount.go [hunk 1]
```diff
@@ -19,6 +19,9 @@ import (
 	"database/sql"
 	"errors"
 	"flag"
+	"github.com/trackit/trackit/aws/usageReports/riEc2"
+	"github.com/trackit/trackit/aws/usageReports/riRdS"
+	onDemandToRiEc2 "github.com/trackit/trackit/onDemandToRI/ec2"
```
Comment (hug33k) [inline server/taskProcessAccount.go]: Wrong input order

File: server/taskElementalProcessAccount.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	const sqlstr = `UPDATE aws_account_elemental_update_job SET
+		completed=?,
+		jobError=?,
+		historyError=?,
+		medialiveChannelsError=?,
+		medialiveInputsError=?,
+		mediaconvertError=?,
+		monthly_reports_generated=?
+	WHERE id=?`
+	_, err := db.Exec(sqlstr, time.Now(), errToStr(jobErr), errToStr(historyErr), medialiveChannelsErr, medialiveInputsErr, mediaconvertErr, historyCreated, updateId)
```
Comment (hug33k) [inline server/taskElementalProcessAccount.go]: Why do you only `errToStr` only on `jobErr` and `historyErr` ?

File: db/migration/0044_add_elemental_process_account_task.sql [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+--   Unless required by applicable law or agreed to in writing, software
+--   distributed under the License is distributed on an "AS IS" BASIS,
+--   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+--   See the License for the specific language governing permissions and
+--   limitations under the License.
+
+CREATE TABLE aws_account_elemental_update_job (
+	id                     INTEGER      NOT NULL AUTO_INCREMENT,
+	created                TIMESTAMP    NOT NULL DEFAULT CURRENT_TIMESTAMP,
+	aws_account_id 				 INTEGER      NOT NULL,
```
Comment (hug33k) [inline db/migration/0044_add_elemental_process_account_task.sql]: Indentation

File: db/migration/0044_add_elemental_process_account_task.sql [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	completed              TIMESTAMP    NOT NULL DEFAULT 0,
+	worker_id              VARCHAR(255) NOT NULL,
+	jobError               VARCHAR(255) NOT NULL DEFAULT "",
+	historyError           VARCHAR(255) NOT NULL DEFAULT "",
+	medialiveChannelsError VARCHAR(255) NOT NULL DEFAULT "",
+	mediaLiveInputsError   VARCHAR(255) NOT NULL DEFAULT "",
```
Comment (hug33k) [inline db/migration/0044_add_elemental_process_account_task.sql]: Why `medialive` then `mediaLive` ?

File: db/migration/0044_add_elemental_process_account_task.sql [hunk 5]
```diff
@@ -0,0 +1,35 @@
+--   Copyright 2017 MSolution.IO
```
Comment (hug33k) [inline db/migration/0044_add_elemental_process_account_task.sql:1]: This migration should be added at the end of `db/schema.sql` also

File: aws/usageReports/history/history.go [hunk 6]
```diff
@@ -245,3 +258,28 @@ func FetchHistoryInfos(ctx context.Context, aa aws.AwsAccount, date time.Time) (
...
+	})
+	complete, err := CheckBillingDataCompleted(ctx, startDate, endDate, aa)
+	if err != nil {
+		return false, err
+	} else if complete == false {
+		logger.Info("Billing data are not completed", nil)
+		return false, ErrBillingDataIncomplete
+	}
+	return getElementalInfos(ctx, aa, startDate, endDate)
+}
```
Comment (hug33k) [inline aws/usageReports/history/history.go]: Missing `\n`

File: aws/usageReports/history/history.go [hunk 7]
```diff
@@ -245,3 +258,28 @@ func FetchHistoryInfos(ctx context.Context, aa aws.AwsAccount, date time.Time) (
...
+func FetchHistoryElementalInfos(ctx context.Context, aa aws.AwsAccount, date time.Time) (bool, error) {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	var startDate, endDate time.Time
+	if date.IsZero() {
+		startDate, endDate = GetHistoryDate()
+	} else {
+		startDate = date
+		endDate = time.Date(date.Year(), date.Month()+1, 0, 23, 59, 59, 999999999, date.Location())
+	}
+	logger.Info("Starting history report", map[string]interface{}{
```
Comment (hug33k) [inline aws/usageReports/history/history.go]: Replace message with `"Starting Elemental history report"`

File: aws/usageReports/mediaconvert/costJobs.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/olivere/elastic"
+
+	"github.com/trackit/jsonlog"
```
Comment (hug33k) [inline aws/usageReports/mediaconvert/costJobs.go:11]: This import should be below `"github.com/olivere/elastic"`

File: aws/usageReports/mediaconvert/costJobs.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+				"error": err,
+			})
+		} else {
+			logger.Error("Query execution failed", map[string]interface{}{"error": err.Error()})
+		}
+		return nil, errors.GetErrorMessage(ctx, err)
+	}
+	logger.Debug("THE RES IS", map[string]interface{}{
+		"RES": res,
+	})
```
Comment (hug33k) [inline aws/usageReports/mediaconvert/costJobs.go]: Please remove debug log

File: aws/usageReports/mediaconvert/costJobs.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+		return nil
+	}
+	err = json.Unmarshal(*res.Aggregations["resourceId"], &response.Id)
+	if err != nil {
+		logger.Error("Unmarshal execution failedd", err)
+		return nil
+	}
+	logger.Debug("response for mediaconvert monthly reports", map[string]interface{}{
+		"response": response.Id,
+	})
```
Comment (hug33k) [inline aws/usageReports/mediaconvert/costJobs.go]: Please remove debug log

File: aws/usageReports/mediaconvert/dailyReport.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/mediaconvert"
+	"github.com/trackit/jsonlog"
+	taws "github.com/trackit/trackit/aws"
```
Comment (hug33k) [inline aws/usageReports/mediaconvert/dailyReport.go:27]: Missing `\n` above this import

File: aws/usageReports/mediaconvert/dailyReport.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+			Role:                 aws.StringValue(job.Role),
+			Status:               aws.StringValue(job.Status),
+			StatusUpdateInterval: aws.StringValue(job.StatusUpdateInterval),
+			Timing: Timing{
+				FinishTime: aws.TimeValue(job.Timing.FinishTime),
+				StartTime:  aws.TimeValue(job.Timing.StartTime),
+				SubmitTime: aws.TimeValue(job.Timing.SubmitTime),
+			},
+			UserMetadata: getUserMetadata(job.UserMetadata),
+			Cost:         0,
```
Comment (hug33k) [inline aws/usageReports/mediaconvert/dailyReport.go:89]: Did you used `gofmt` for indentation ?

File: aws/usageReports/mediaconvert/monthlyReport.go [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package mediaconvert
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/trackit/trackit/es"
```
Comment (hug33k) [inline aws/usageReports/mediaconvert/monthlyReport.go]: Wrong import order

File: aws/usageReports/mediaconvert/monthlyReport.go [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+		"endDate":      endDate.Format("2006-01-02T15:04:05Z"),
+	})
+	costs := getMediaConvertJobCosts(ctx, aa, startDate, endDate)
+	/*already, err := utils.CheckMonthlyReportExists(ctx, startDate, aa, IndexPrefixMediaConvertReport)
+	if err != nil {
+		return false, err
+	} else if already {
+		logger.Info("There is already an MediaConvert monthly report", nil)
+		return false, nil
+	}*/
```
Comment (hug33k) [inline aws/usageReports/mediaconvert/monthlyReport.go]: 🤔

File: aws/usageReports/mediaconvert/utils.go [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+		JobTemplate          string              `json:"jobTemplate"`
+		OutputGroupDetails   []OutputGroupDetail `json:"outputGroupDetails"`
+		Queue                string              `json:"queue"`
+		RetryCount           int64               `json:"retryCount"`
+		Role                 string              `json:"role"`
+		Status               string              `json:"status"`
+		StatusUpdateInterval string              `json:"statusUpdateInterval"`
+		Timing               Timing
+		UserMetadata         map[string]string `json:"userMetadata"`
+		Cost                 float64           `json:"cost"`
```
Comment (hug33k) [inline aws/usageReports/mediaconvert/utils.go]: Indentation

File: aws/usageReports/medialive/channelCosts.go [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package medialive
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
```
Comment (hug33k) [inline aws/usageReports/medialive/channelCosts.go]: Wrong import order

File: aws/usageReports/medialive/channelCosts.go [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+		Arn    string
+	}
+)
+
+//getElasticSearchCost get Elemental MediaLive resources from lineitems in ES to get jobs costs
+func getElasticSearchCost(ctx context.Context, startDate, endDate time.Time, userId int) (*elastic.SearchResult, error) {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	query := elastic.NewBoolQuery()
+	now := time.Now().UTC()                                                                 // line to remove
+	startDate = time.Date(now.Year()-1, now.Month()-1, 1, 0, 0, 0, 0, now.Location()).UTC() // line to remove
```
Comment (hug33k) [inline aws/usageReports/medialive/channelCosts.go]: 🤔

File: aws/usageReports/medialive/channelCosts.go [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+func getMediaLiveChannelCosts(ctx context.Context, aa taws.AwsAccount, startDate, endDate time.Time) []ChannelInformations {
+	var response ResponseChannelIdCostMonthly
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	res, err := getElasticSearchCost(ctx, startDate, endDate, aa.UserId)
+	if err != nil {
+		return nil
+	}
+	fmt.Printf("response = %v\n", res)
```
Comment (hug33k) [inline aws/usageReports/medialive/channelCosts.go]: Please remove debug print

File: aws/usageReports/medialive/dailyReport.go [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getInputsFromAWS(inputChan chan Input, svc *medialive.MediaLive, region string, token *string) (*string, error) {
+	listInput, err := svc.ListInputs(&medialive.ListInputsInput{NextToken: token})
+	fmt.Printf("listInput = %v\n", listInput)
```
Comment (hug33k) [inline aws/usageReports/medialive/dailyReport.go]: Please remove debug log

File: aws/usageReports/medialive/dailyReport.go [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+			return err
+		}
+	}
+	return nil
+}
+
+func getChannelsFromAWS(channelChan chan Channel, svc *medialive.MediaLive, region string, token *string) (*string, error) {
+	listChannel, err := svc.ListChannels(&medialive.ListChannelsInput{NextToken: token})
+	if err != nil {
+		fmt.Printf("Error when listing channels for medialive: %v\n", err)
```
Comment (hug33k) [inline aws/usageReports/medialive/dailyReport.go]: Please remove debug print

File: aws/usageReports/medialive/dailyReport.go [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+		}
+	}
+	return nil
+}
+
+func getInputsFromAWS(inputChan chan Input, svc *medialive.MediaLive, region string, token *string) (*string, error) {
+	listInput, err := svc.ListInputs(&medialive.ListInputsInput{NextToken: token})
+	fmt.Printf("listInput = %v\n", listInput)
+	if err != nil {
+		fmt.Printf("error input medialive = %v\n", err)
```
Comment (hug33k) [inline aws/usageReports/medialive/dailyReport.go]: Please remove debug log

File: aws/usageReports/medialive/fetchChannelsInputs.go [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+	return channelsList
+}
+
+func fetchInputs(ctx context.Context, aa taws.AwsAccount, startDate, endDate time.Time, regions []string, account string,
+	creds *credentials.Credentials) []InputReport {
+	inputsCosts := getMediaLiveInputCosts(ctx, aa, startDate, endDate)
+	inputChans := make([]<-chan Input, 0, len(regions))
+	for _, cost := range inputsCosts {
+		for _, region := range regions {
+			log.Printf("REGION = %v\n", region)
```
Comment (hug33k) [inline aws/usageReports/medialive/fetchChannelsInputs.go]: Please remove debug log

File: aws/usageReports/medialive/mappingsChannels.go [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+							"type": "nested",
+							"properties": {
+								"key": {
+									"type": "keyword"
+								},
+								"value": {
+									"type": "keyword"
+								}
+							}
+						},
```
Comment (hug33k) [inline aws/usageReports/medialive/mappingsChannels.go]: 🤔

File: aws/usageReports/medialive/mappingsInput.go [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+							"type": "nested",
+							"properties": {
+								"key": {
+									"type": "keyword"
+								},
+								"value": {
+									"type": "keyword"
+								}
+							}
+						},
```
Comment (hug33k) [inline aws/usageReports/medialive/mappingsInput.go]: 🤔

File: aws/usageReports/medialive/monthlyReport.go [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+	channels, err := fetchMonthlyChannelsStats(ctx, aa, startDate, endDate)
+	if err != nil {
+		return false, err
+	}
+	err = importChannelsToEs(ctx, aa, channels)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
```
Comment (hug33k) [inline aws/usageReports/medialive/monthlyReport.go]: Missing `\n`

File: aws/usageReports/medialive/monthlyReport.go [hunk 26]
```diff
@@ -0,0 +1,293 @@
...
+		PipelinesRunningCount: aws.Int64Value(describeChannel.PipelinesRunningCount),
+		State:                 aws.StringValue(describeChannel.State),
+		Tags:                  getChannelTags(describeChannel.Tags),
+		Cost:                  cost.Cost,
+	}
+	return nil
+}
+
+// fetchMonthlyInput sends in channelInfoChan the channels fetched from DescribeChannels
+// and filled by DescribeChannels and getChannelStats.
```
Comment (hug33k) [inline aws/usageReports/medialive/monthlyReport.go]: 🤔

**Review by hug33k — CHANGES_REQUESTED** [review]

File: aws/usageReports/medialive/mappingsChannels.go (hunk 23 above)
Comment (jgchoppe) [reply aws/usageReports/medialive/mappingsChannels.go]: @hug33k I don't see the issue ?

File: aws/usageReports/mediaconvert/dailyReport.go (hunk 12 above)
Comment (jgchoppe) [reply aws/usageReports/mediaconvert/dailyReport.go:89]: yes

**Comment by alexandre-snr** [comment]
This task was related to a project which was cancelled in favor of Tagbot.

**Pull request closed without merging**